pallets/dapp-staking/src/types.rs

```diff
@@ -993,35 +1001,98 @@ impl EraInfo {
+    fn default() -> Self {
+        let max = MaxBonusMoves::get();
+        BonusStatus::SafeMovesRemaining(max)
+    }
+}
+
+impl<MaxBonusMoves: Get<u8>> BonusStatus<MaxBonusMoves> {
+    /// Decrease the number of remaining safe moves by 1.
+    /// If the counter reaches 0, the bonus is forfeited.
+    pub fn decrease_moves(&mut self) {
+        *self = match self {
+            BonusStatus::SafeMovesRemaining(counter) => {
+                if *counter == 0 {
+                    BonusStatus::BonusForfeited
+                } else {
+                    BonusStatus::SafeMovesRemaining(counter.saturating_sub(1))
+                }
+            }
+            _ => BonusStatus::BonusForfeited,
+        }
+    }
+
+    /// Check if bonus rewards are preserved.
+    pub fn has_bonus(&self) -> bool {
+        matches!(self, BonusStatus::SafeMovesRemaining(_))
+    }
+}
+
+impl<MaxBonusMoves: Get<u8>> PartialEq for BonusStatus<MaxBonusMoves> {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (BonusStatus::BonusForfeited, BonusStatus::BonusForfeited) => true,
+            (BonusStatus::SafeMovesRemaining(a), BonusStatus::SafeMovesRemaining(b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+impl<MaxBonusMoves: Get<u8>> Eq for BonusStatus<MaxBonusMoves> {}
+
 /// Information about how much a particular staker staked on a particular smart contract.
 ///
 /// Keeps track of amount staked in the 'voting subperiod', as well as 'build&earn subperiod'.
-#[derive(Encode, Decode, MaxEncodedLen, Copy, Clone, Debug, PartialEq, Eq, TypeInfo, Default)]
-pub struct SingularStakingInfo {
+#[derive(Encode, MaxEncodedLen, Copy, Clone, Debug, TypeInfo, Default)]
+#[scale_info(skip_type_params(MaxBonusMoves))]
+pub struct SingularStakingInfo<MaxBonusMoves: Get<u8>> {
     /// Amount staked before, if anything.
     pub(crate) previous_staked: StakeAmount,
     /// Staked amount
     pub(crate) staked: StakeAmount,
-    /// Indicates whether a staker is a loyal staker or not.
-    pub(crate) loyal_staker: bool,
+    /// Tracks the remaining allowable actions to preserve a staking bonus, based on the staker's behavior during specific subperiods.
+    pub(crate) bonus_status: BonusStatus<MaxBonusMoves>,
 }
 
-impl SingularStakingInfo {
+impl<MaxBonusMoves: Get<u8>> SingularStakingInfo<MaxBonusMoves> {
     /// Creates new instance of the struct.
     ///
     /// ## Args
     ///
     /// `period` - period number for which this entry is relevant.
     /// `subperiod` - subperiod during which this entry is created.
     pub(crate) fn new(period: PeriodNumber, subperiod: Subperiod) -> Self {
+        // Bonus staking is only possible if stake is first made during the voting subperiod.
+        let bonus_status = if subperiod == Subperiod::Voting {
+            BonusStatus::<MaxBonusMoves>::default()
+        } else {
+            BonusStatus::<MaxBonusMoves>::BonusForfeited
+        };
```

Comment:
If this has to be decided at this point - it might be better to provide the value from the outside.

Current approach seems hacky & incomplete.